lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -819,30 +838,51 @@ CreateSiFiveMetadataPass::emitSitestBlackboxMetadata(ObjectModelIR &omir) {
   // Find all extmodules in the circuit. Check if they are black-listed from
   // being included in the list. If they are not, separate them into two
   // groups depending on if theyre in the DUT or the test harness.
-  SmallVector<StringRef> dutModules;
-  SmallVector<StringRef> testModules;
+  SmallVector<StringRef> dutLibs;
+  SmallVector<StringRef> testLibs;
+
   for (auto extModule : circuitOp.getBodyBlock()->getOps<FExtModuleOp>()) {
+    SmallVector<StringRef> libs;
+
     // If the module doesn't have a defname, then we can't record it properly.
     // Just skip it.
     if (!extModule.getDefname())
       continue;
 
-    // If its a generated blackbox, skip it.
     AnnotationSet annos(extModule);
-    if (llvm::any_of(blackListedAnnos, [&](auto blackListedAnno) {
-          return annos.hasAnnotation(blackListedAnno);
-        }))
-      continue;
 
-    // Record the defname of the module.
-    bool inDut = false;
-    if (instanceInfo->anyInstanceInEffectiveDesign(extModule)) {
-      inDut = true;
-      dutModules.push_back(*extModule.getDefname());
-    } else {
-      testModules.push_back(*extModule.getDefname());
+    bool isBlacklistedBlackbox = llvm::any_of(blackListedAnnos, [&](auto blackListedAnno) {
+      return annos.hasAnnotation(blackListedAnno);
+    });
+
+    if (!isBlacklistedBlackbox)
+      libs.push_back(*extModule.getDefname());
+
+    bool hasLibs = false;
+    if (auto libsAnno = annos.getAnnotation(sitestBlackBoxLibrariesAnnoClass)) {
+      if (auto libsAttr = libsAnno.getMember<ArrayAttr>("libraries")) {
+        for (auto lib : libsAttr) {
+          if (auto libStr = dyn_cast<StringAttr>(lib)) {
+            libs.push_back(libStr.getValue());
+            hasLibs = true;
+          }
+        }
+      }
+    }
+
+    if (!libs.empty()) {
```

Comment:
nit: I think our [preferred style](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code) would be more like

```
if (libs.empty())
  continue;
```

And to generally pull early returns / continues as soon as possible. I don't think this could be pulled any earlier, but that is the general gist.